### PR TITLE
feat: readarr

### PR DESCRIPTION
**Pull request by flying-sausages**
- feat: readarr installer
- test: add test definition for readarr
- style: make the nginx config legible

## Description

Add readarr

## Fixes issues: 
N/A
## Proposed Changes:
New app with tests and all

## Change Categories

- New feature 

## Checklist

- [x] Branch was made off the `develop` branch and the PR is targetting the `develop` branch
- [x] Docs have been made OR are not necessary
    - PR link: https://github.com/liaralabs/docs.swizzin.ltd/pull/43
- [x] Changes to panel have been made OR are not necessary
    - PR link: https://github.com/swizzin/swizzin_dashboard/pull/24
- [x] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Code conforms to project structure [(See more)](https://swizzin.ltd/dev/structure)
- [x] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Testing was done
   - [x] Tests created or no new tests necessary
   - [x] Tests executed

### Architectures

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Focal 		|			|			|	yup		|				|
| Bionic		|			|			|			|				|
| Buster		|			|			|			|				|
| Stretch		|			|			|			|				|
| Raspbian  	|	⚫️		|			|	⚫️		|	⚫️			|

### CODE REVIEW

**Comment by flying-sausages** [comment]
allegedly a develop version is around the corner, so let's wait on that first.

**Comment by insomniac-afk** [comment]
Prowlarr is currently using port 9696 as well just an fyi, Making a note here as Liara suggested.

Prowlarr port here: https://github.com/swizzin/swizzin/blob/001bd702300e533ec7213911cc20428732239867/scripts/install/prowlarr.sh#L24

**Comment by Aethaeran** [comment]
https://github.com/swizzin/swizzin/blob/feat/readarr/scripts/nginx/readarr.sh
Needs `app_port` to be changed to match the changes last week in the readarr install script (change from 9696 to 8787)
Needs `app_sslport` the be changed to _**NOT**_ match Prowlarr's ssl port.

https://github.com/swizzin/swizzin/blob/feat/readarr/scripts/install/readarr.sh
Needs `app_sslport` added so that it can pass it to Readarr 's config.xml on installation. (a different one than Prowlarr is using)

**Review by brettpetch — COMMENTED** [review]
lets wait on this until we change from nightly to develop ;)

> 
> 
> Prowlarr is currently using port 9696 as well just an fyi, Making a note here as Liara suggested.
> 
> Prowlarr port here:
> 
> https://github.com/swizzin/swizzin/blob/001bd702300e533ec7213911cc20428732239867/scripts/install/prowlarr.sh#L24



> 
> 
> https://github.com/swizzin/swizzin/blob/feat/readarr/scripts/nginx/readarr.sh
> Needs `app_port` to be changed to match the changes last week in the readarr install script (change from 9696 to 8787)
> Needs `app_sslport` the be changed to _**NOT**_ match Prowlarr's ssl port. (6969)
> 
> https://github.com/swizzin/swizzin/blob/feat/readarr/scripts/install/readarr.sh
> Needs `app_sslport` added so that it can pass it to Readarr 's config.xml on installation. (a different one than Prowlarr is using)

solved now.

**Comment by Aethaeran** [comment]
```
# sudo box test readarr
INFO    Running test for readarr
...     Checking readarr service is active
        ✔   Done
...     Checking if port 8787 is reachable via curl
        ✔   Done
...     Checking if readarr is reachable via nginx
        ✔   Done
SUCCESS No problems were encountered
```

**Comment by brettpetch** [comment]
Thx. 

/b

**Comment by she11sh0cked** [comment]
Did an install on my seedbox.
https://she11sh0cked.tk/readarr is reachable, but doesn't show up on the dashboard.

Here is the install output:
```
ubuntu@seedbox:/etc/swizzin$ sudo box
INFO    Installing readarr
...     Performing installation of 2 apt packages
        (curl sqlite3)
        ✔   Apt install complete
...     Downloading release archive
        ✔   Archive downloaded
...     Extracting archive
        ✔   Archive extracted
...     Configuring nginx
grep: /home/she11sh0cked/.config/Readarr/config.xml: No such file or directory
        ✔   Nginx configured
...     Installing Systemd service
        ✔   Readarr service installed and enabled
SUCCESS Readarr installed
```

Don't know about the error... The config file exists:
```
ubuntu@seedbox:/etc/swizzin$ sudo cat /home/she11sh0cked/.config/Readarr/config.xml
<Config>
  <LogLevel>debug</LogLevel>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <BindAddress>127.0.0.1</BindAddress>
  <Port>8787</Port>
  <SslPort>6868</SslPort>
  <EnableSsl>False</EnableSsl>
  <LaunchBrowser>False</LaunchBrowser>
  <ApiKey></ApiKey>
  <AuthenticationMethod>None</AuthenticationMethod>
  <UrlBase>readarr</UrlBase>
  <Branch>nightly</Branch>
</Config>
```

Lastly, here are my specs:
```
ubuntu@seedbox:/etc/swizzin$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.3 LTS
Release:        20.04
Codename:       focal

ubuntu@seedbox:/etc/swizzin$ uname -a
Linux seedbox 5.4.0-89-generic #100-Ubuntu SMP Fri Sep 24 14:50:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by bakerboy448** [comment]
@she11sh0cked Readarr never actually started/ran.  See the wiki for troubleshooting steps.  You can also try starting it ./opt/Readarr/Readarr and see what the console complains about

File: scripts/install/readarr.sh [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+user="$READARR_OWNER"
+swiz_configdir="/home/$user/.config"
+app_configdir="$swiz_configdir/${app_name^}"
+app_group="$user"
+app_port="8787"
+app_reqs=("curl" "sqlite3")
+app_servicefile="$app_name.service"
+app_dir="/opt/${app_name^}"
+app_binary="${app_name^}"
+app_lockname="${app_name//-/}"
```
Comment (bakerboy448) [inline scripts/install/readarr.sh:28]: Change to develop if this will not be merged prior

File: scripts/nginx/readarr.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+app_sslport="6868"
+app_baseurl="$app_name"
+app_configdir="/home/$user/.config/${app_name^}"
+app_servicefile="${app_name}.service"
+app_branch="nightly"
+
+cat > /etc/nginx/apps/$app_name.conf << ARRNGINX
+location /$app_baseurl {
+    proxy_pass          http://127.0.0.1:$app_port/$app_baseurl;
+    proxy_set_header    Host                \$proxy_host;
```
Comment (bakerboy448) [inline scripts/nginx/readarr.sh:23]: $host

Comment (bakerboy448) [reply scripts/nginx/readarr.sh:23]: Missing header `proxy_set_header X-Forwarded-Host $host;`

File: scripts/nginx/readarr.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    proxy_pass      http://127.0.0.1:$app_port/$app_baseurl/api;
+}
+
+# Allow Content
+location /$app_baseurl/Content {
+    auth_request    off;
+    proxy_pass      http://127.0.0.1:$app_port/$app_baseurl/Content;
+}
+
+# Allow Indexers  $1 matches the regex
```
Comment (bakerboy448) [inline scripts/nginx/readarr.sh:48]: Why is this here? This has nothin to do with readarr, remove this block

**Review by bakerboy448 — CHANGES_REQUESTED** [review]

**Comment by she11sh0cked** [comment]
> 
> 
> @she11sh0cked Readarr never actually started/ran. See the wiki for troubleshooting steps. You can also try starting it ./opt/Readarr/Readarr and see what the console complains about

It did start though. It just didn't show up in the dashboard. (Running great btw)

**Comment by bakerboy448** [comment]
No longer blocked externally

however this PR needs changes
https://github.com/Readarr/Readarr/issues/666#issuecomment-984690118

**Comment by userdocs** [comment]
I take a look at it over the weekend.

File: scripts/install/readarr.sh [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+app_branch="nightly" # Change to develop/stable when available, tell users to migrate manually.
```
Comment (bakerboy448) [inline scripts/install/readarr.sh:29]: develop not nightly

File: scripts/nginx/readarr.sh [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+if ! READARR_OWNER="$(swizdb get $app_name/owner)"; then
+    READARR_OWNER=$(_get_master_username)
+fi
+user="$READARR_OWNER"
+
+app_port="8787"
+app_sslport="7979"
+app_baseurl="$app_name"
+app_configdir="/home/$user/.config/${app_name^}"
+app_servicefile="${app_name}.service"
```
Comment (bakerboy448) [inline scripts/nginx/readarr.sh:17]: develop not nightly

**Review by bakerboy448 — CHANGES_REQUESTED** [review]
Readarr develop has dropped and now exists

there is no longer an external dependency issue and this all needs to be moved to develop from nightly


Also did kebpubnotify whatever get fixed for calibre? Didn't see the PR but swear it was

File: scripts/nginx/readarr.sh [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    proxy_set_header    Connection  \$http_connection;
+}
+
+# Allow the App API
+location /$app_baseurl/api {
+    auth_request    off;
+    proxy_pass      http://127.0.0.1:$app_port/$app_baseurl/api;
+}
+
+# Allow Content
```
Comment (bakerboy448) [inline scripts/nginx/readarr.sh:43]: Don't need content

**Review by bakerboy448 — CHANGES_REQUESTED** [review]

File: scripts/nginx/readarr.sh (hunk 6 above)
Comment (brettpetch) [reply scripts/nginx/readarr.sh:43]: content is still enabled in the official documentation.


```nginx
location /readarr {
  proxy_pass        http://127.0.0.1:8787/readarr;
  proxy_set_header Host $host;
  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header X-Forwarded-Host $host;
  proxy_set_header X-Forwarded-Proto https;
  proxy_redirect off;

  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection $http_connection;
}
  location /readarr/api {
  auth_request off;
  proxy_pass       http://127.0.0.1:8787/readarr/api;
}

  location /readarr/Content { 
    auth_request off;
    proxy_pass http://127.0.0.1:8787/readarr/Content;
 }
```

Comment (bakerboy448) [reply scripts/nginx/readarr.sh:43]: official docs updated  to match other \*Arrs

```nginx
location /readarr {
    proxy_pass http://127.0.0.1:8787;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
}
# Allow the API External Access via NGINX
location ~ /readarr/api {
    auth_request off;
    proxy_pass http://127.0.0.1:8787;
}
```

**Comment by brettpetch** [comment]
Install on OCI ARM64 instance
```
INFO    Currently on branch: feat/readarr
INFO    Installing readarr
DOCS    Further reference: https://swizzin.ltd/applications/readarr
...     Downloading release archive
        ✔   Archive downloaded
...     Extracting archive
        ✔   Archive extracted
...     Installing Systemd service
        ✔   Readarr service installed and enabled
...     Readarr is loading...
        ✔   Loading finished
...     Configuring nginx
        ✔   Nginx configured
SUCCESS Readarr installed
```
Test on same instance.
```
[Wed Feb 23 03:08:01] LOG       Running test for readarr
[Wed Feb 23 03:08:01] LOG       Checking readarr service is active
[Wed Feb 23 03:08:01] LOG       Checking if port 8787 is reachable via curl
[Wed Feb 23 03:08:01] LOG       Checking if readarr is reachable via nginx
[Wed Feb 23 03:08:01] LOG        passed all tests
[Wed Feb 23 03:08:01] SUCCESS   All tests completed succesfully.
Test log can be found in /root/logs/tests-2022.02.23-03:08:01.txt
```

**Comment by she11sh0cked** [comment]
@brettpetch I take it this still needs testing? Or what is keeping it from getting merged? :)

**Comment by she11sh0cked** [comment]
OS info:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye

$ uname -a
Linux eva-server 5.10.0-10-amd64 #1 SMP Debian 5.10.84-1 (2021-12-08) x86_64 GNU/Linux
```
Installation:
```
$ sudo box
INFO    Currently on branch: feat/readarr
INFO    Installing readarr
...     Downloading release archive
        ✔   Archive downloaded
...     Extracting archive
        ✔   Archive extracted
...     Installing Systemd service
        ✔   Readarr service installed and enabled
...     Readarr is loading...
        ✔   Loading finished
...     Configuring nginx
        ✔   Nginx configured
SUCCESS Readarr installed
```
Testing:
```
$ sudo box test readarr
INFO    Currently on branch: feat/readarr
SUCCESS All tests completed succesfully.
        Test log can be found in /root/logs/tests-2022.03.11-15:31:51.txt

$ sudo cat /root/logs/tests-2022.03.11-15:31:51.txt
[Fri Mar 11 15:31:51] LOG       Running test for readarr
[Fri Mar 11 15:31:51] LOG       Checking readarr service is active
[Fri Mar 11 15:31:51] LOG       Checking if port 8787 is reachable via curl
[Fri Mar 11 15:31:52] LOG       Checking if readarr is reachable via nginx
[Fri Mar 11 15:31:52] LOG        passed all tests
[Fri Mar 11 15:31:52] SUCCESS   All tests completed succesfully.
Test log can be found in /root/logs/tests-2022.03.11-15:31:51.txt
```

**Comment by brettpetch** [comment]
> OS info:
> 
> ```
> $ lsb_release -a
> No LSB modules are available.
> Distributor ID: Debian
> Description:    Debian GNU/Linux 11 (bullseye)
> Release:        11
> Codename:       bullseye
> 
> $ uname -a
> Linux eva-server 5.10.0-10-amd64 #1 SMP Debian 5.10.84-1 (2021-12-08) x86_64 GNU/Linux
> ```
> 
> Installation:
> 
> ```
> $ sudo box
> INFO    Currently on branch: feat/readarr
> INFO    Installing readarr
> ...     Downloading release archive
>         ✔   Archive downloaded
> ...     Extracting archive
>         ✔   Archive extracted
> ...     Installing Systemd service
>         ✔   Readarr service installed and enabled
> ...     Readarr is loading...
>         ✔   Loading finished
> ...     Configuring nginx
>         ✔   Nginx configured
> SUCCESS Readarr installed
> ```
> 
> Testing:
> 
> ```
> $ sudo box test readarr
> INFO    Currently on branch: feat/readarr
> SUCCESS All tests completed succesfully.
>         Test log can be found in /root/logs/tests-2022.03.11-15:31:51.txt
> 
> $ sudo cat /root/logs/tests-2022.03.11-15:31:51.txt
> [Fri Mar 11 15:31:51] LOG       Running test for readarr
> [Fri Mar 11 15:31:51] LOG       Checking readarr service is active
> [Fri Mar 11 15:31:51] LOG       Checking if port 8787 is reachable via curl
> [Fri Mar 11 15:31:52] LOG       Checking if readarr is reachable via nginx
> [Fri Mar 11 15:31:52] LOG        passed all tests
> [Fri Mar 11 15:31:52] SUCCESS   All tests completed succesfully.
> Test log can be found in /root/logs/tests-2022.03.11-15:31:51.txt
> ```

I don’t have merge perms. I have asked @liaralabs for a final review.

**Review by brettpetch — APPROVED** [review]
